hdfs support

skewballfox opened the pull request:
Still trying to figure a few things out. This might be easier than I was initially expecting given it's the default shared filesystem other implementations have to be compatible with. 


Some relevant info I've found so far:
- [spark docs discussing the hdfs state store provider](https://spark.apache.org/docs/3.5.2/structured-streaming-programming-guide.html#hdfs-state-store-provider). Note there were some other hits in it's RDD programming guide
- since hdfs is only mentioned in the spark docs in reference to streaming datasets, I think [these might be the relevant test](https://github.com/search?q=repo%3Aapache%2Fspark+path%3A%2F%5Estreaming%5C%2Fsrc%5C%2Ftest%5C%2Fscala%5C%2Forg%5C%2Fapache%5C%2Fspark%5C%2Fstreaming%5C%2F%2F++hdfs&type=code). I looked through the rest of the hits not limited to that directory, and I didn't find any other test that reference hdfs directly.
- [Notes on unit testing spark jobs that require HDFS](https://zedar.gitbooks.io/spark-hadoop-notes/content/spark_unit_testing_with_hdfs.html)

all hits for `object_store` in `crates` happen within sail plan, so I'm guessing this should primarily be implemented within that crate?

skewballfox commented:
hey, trying to setup spark for test but I'm running into an issue. even with a clean slate I'm getting a error somewhere 

```
rm -rf opt/*
git clone https://github.com/apache/spark.git opt/spark
git clone https://github.com/ibis-project/testing-data.git opt/ibis-testing-data
scripts/spark-tests/build-pyspark.sh
```
here's the last bit of output. scrolling up can't take me  past a long stream of "copying " statements to the origin.
```
...
copying pyspark/tests/typing/test_resultiterable.yml -> pyspark-3.5.1/pyspark/tests/typing
copying pyspark.egg-info/SOURCES.txt -> pyspark-3.5.1/pyspark.egg-info
Writing pyspark-3.5.1/setup.cfg
creating dist
Creating tar archive
removing 'pyspark-3.5.1' (and everything under it)
error: Your local changes to the following files would be overwritten by checkout:
	pom.xml
Please commit your changes or stash them before you switch branches.
Aborting
```
given [this post](https://stackoverflow.com/questions/45168408/creating-tar-gz-in-dist-folder-with-python-setup-py-install) has a match for all the non-specific strings, I'm guessing this error comes from setup.py or immediately after setup.py is called. Any ideas what's going on?

skewballfox commented:
Would it be alright if I pushed some "off-topic" commits? I added some changes to the `build-pyspark.sh` script that make it easier to resume after a failed partially-completed run, and makes sure the venv is setup/sourced before running the python command

linhr commented:
I just took a look at the script change and I feel it goes a bit beyond what the script is supposed to do. Since the script is also used in CI, it may conflict with other parts of the GitHub Actions workflow.

Would you mind reverting the script change? Sorry for the back and forth. These are valuable suggested changes though. We'll definitely consider a way to incorporate the idea to make the developer experience better.

linhr commented:
BTW I like the idea of using a lock file to skip expensive Maven build. I'll need some time to test it in CI so let's revisit this in a future PR.

skewballfox commented:
> I just took a look at the script change and I feel it goes a bit beyond what the script is supposed to do. Since the script is also used in CI, it may conflict with other parts of the GitHub Actions workflow.
> 
> Would you mind reverting the script change? Sorry for the back and forth. These are valuable suggested changes though. We'll definitely consider a way to incorporate the idea to make the developer experience better.

no worries, I used git reset to undo the last commit. and I'll push the script to a separate branch. Sorry for the delayed response, been running into issues getting the test environment setup

skewballfox commented:
a couple things. 

Is there a way to get pyo3 to use a uv installed python version? I tried running `uv python install 3.11` prior to running `start_server.sh` but would still get linker errors for python3.11. I fixed this by installing it system level (`sudo dnf install python3.11-devel`), but was wondering if there was a way to easily avoid this (I try to keep development dependencies user level)

Second, I'm having trouble getting the `run-client.sh` script to work after everything has been set up. here's the output:
```
user@fedora ~/t/sail (hdfs_support) [1]> scripts/spark-tests/run-tests.sh
Removing existing test logs...
Test suite: test-connect
ERROR: module or package not found: pyspark.sql.tests.connect (missing __init__.py?)

================================================= test session starts ==================================================
platform linux -- Python 3.11.9, pytest-8.3.3, pluggy-1.5.0
rootdir: /path/to/sail
configfile: pyproject.toml
plugins: hypothesis-6.112.1, xdist-3.6.1, timeout-2.3.1, snapshot-0.9.0, reportlog-0.4.0, repeat-0.9.3, mock-3.14.0, pytest_httpserver-1.1.0
collected 0 items

-------- generated report log file: /path/to/sail/tmp/spark-tests/latest/test-connect.jsonl ---------
================================================ no tests ran in 0.17s =================================================
Test suite: doctest-column
===================== test session starts ======================
platform linux -- Python 3.11.9, pytest-8.3.3, pluggy-1.5.0
rootdir: /path/to/sail
configfile: pyproject.toml
plugins: hypothesis-6.112.1, xdist-3.6.1, timeout-2.3.1, snapshot-0.9.0, reportlog-0.4.0, repeat-0.9.3, mock-3.14.0, pytest_httpserver-1.1.0
collected 33 items

.venvs/test/lib/python3.11/site-packages/pyspark/sql/column.py E [  3%]
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE                         [100%]

- generated report log file: /path/to/sail/tmp/spark-tests/latest/doctest-column.jsonl -
====================== 33 errors in 0.22s ======================
```
I've made sure `hatch run test:install-pyspark` has run, and `run-server.sh` is running on a separate terminal. Kinda scratching my head on this one, it looks as though the issue comes from pyspark. I tried cding into `opt/spark/python`, running `python setup.py sdist` again, and rerunning the install script with hatch, but error is the same. Is there some step I'm missing?

skewballfox commented:
> Although, i'm also curious if maybe there is a problem with the Java installation. Unfortunately running the Spark tests requires Java to be installed.

I have java installed and `JAVA_HOME` set to `/usr/lib/jvm/java-17-openjdk`

> Can we try to get some more verbose log outputs? It seems like doctest-column is able to run tests but they all fail. Maybe there are some useful error logs there:

when I ran that command. I'm getting a series of no such file or directory errors for files such as `'/path/to/sail/.venvs/test/lib/python3.11/site-packages/pyspark/python/test_support'`. when I look in the parent directory mentioned, the only contents are `.venvs/test/lib/python3.11/site-packages/pyspark/python/lib/pyspark.zip`, but when listing the contents of the zip, the missing files (i.e. `test_support`) are there. I'm guessing some late stage build step is suppose to unpack that file but isn't?

linhr commented:
Could you run the Spark build script unmodified and see if it works? I took another look at #200 and it seems to me that the patch lock file may cause the patch to be skipped when you build the Python package. If the patch is not applied, neither the tests nor the test support files will be in the correct location in the zip package.

The Maven lock file seems fine to me though.

skewballfox commented:
I renamed my current copy of sail and recloned the repo. Suprisingly it worked, but I'm not sure why given I removed the opt directory a few times (including the lock files), so I retested on my original copy (pruned/recreated the virtual envs, removed opt, reran the build script, etc). It's working now

I think I may finally have a hadoop container setup in pseudo-distributed mode for testing hdfs support. Where should it go? under scripts, opts or somewhere else?

linhr commented:
Glad to hear that you got the setup working!

Let's define the container in `compose.yml`.

Testing external systems with containers is not part of CI right now but it could be future work. For this PR, it would be good enough to have a local setup. Either Python or Rust tests would be great.

skewballfox commented:
hey, kind of stuck atm. I'm setting up hdfs in a local container, verifying it's working via the web portal, and then connecting to the hdfs container using the method mentioned in the dev docs.
```
p/t/sail> env SPARK_CONNECT_MODE_ENABLED=1 SPARK_REMOTE="sc://localhost:50051" hatch run pyspark
>>>path=hdfs://localhost:9000/user/skewballfox/test.json
>>>spark.read.json(path).show()
```
This throws a Object store error originating from this  [generic IO error](https://github.com/Kimahriman/hdfs-native/blob/ebb5a2f98ab9fe487091403f45324650e3cd65fb/rust/src/error.rs#L9). I know it's connecting since the client returns okay, and because if I change it to a non-existent file, I get an error indicating the resource doesn't exist. So it's making it somewhere past the object store instantiation and failing during execution (I think). I'm having trouble tracing it any further than that though.

here's the last bit of traceback from python:
```
File "/p/t/sail/.venvs/default/lib/python3.11/site-packages/pyspark/sql/connect/client/core.py", line 1503, in _handle_error
    self._handle_rpc_error(error)
  File "p/t/sail/.venvs/default/lib/python3.11/site-packages/pyspark/sql/connect/client/core.py", line 1539, in _handle_rpc_error
    raise convert_exception(info, status.message) from None
pyspark.errors.exceptions.connect.SparkRuntimeException: Object Store error: Generic HdfsObjectStore error: IO error occurred while communicating with HDFS
```
here's the last trace from the server
```
[2024-09-19T18:39:08Z DEBUG sail_spark_connect::server trace: 241683380366876879727100582833342366091 span: 5329696312847761413] ReleaseExecuteResponse { session_id: "487e2373-2881-41e9-ad87-d8362c5ef938", operation_id: Some("4e1997e1-2bd1-433a-a252-1826cb68986e") }
```
which is returned by `release_execute`

any ideas what the issue is or have suggestions on what I should try to track down the problem?

skewballfox commented:
Just an update, Managed to finally read a json file from the hdfs instance, was a configuration issue. Still failing to write to hdfs, though that is probably also a configuration issue (working on fixing it).

Launching the hdfs container still doesn't work locally via podman compose. It keeps dying after around 15 seconds without producing any error and returning a normal exit code. I'm running the docker file under `scripts/hadoop` via
```sh
env HADOOP_USER_NAME=sail podman build -t hadoop-container .
podman run -it -p 50070:9870 -p 9000:9000 -p 9864:9864 -p 9866:9866 hadoop-container /bin/bash
#/hdfs-init.sh
```

I think i might have found a bug in the s3 code that may be upstream. If you upload a json file via the web portal, and then try to read it via `spark.read.json(path).show()` you'll get: `Json error: Not valid JSON: EOF while parsing an object at line 1 column 1`. It's definitely valid json, tested with a few files uploaded to the minIO instance

This doesn't happen with the code from the docs, but the query `spark.sql("SELECT 1").write.json(path)` isn't doing what it looks like. `bar.json` is created as a directory, and json files are created under it with the write command. When you look at the metadata for those files the content-type is listed as `binary/octet-stream`, whereas the files it fails to read have the content-type `application/json`. you can run read json commands on the octet-stream files created without issue

linhr commented:
> bar.json is created as a directory, and json files are created under it with the write command.

This is the expected behavior. The content type needs investigation though.

skewballfox commented:
related issue for [hdf-native-object-store](https://github.com/datafusion-contrib/hdfs-native-object-store/issues/9)

skewballfox commented:
So I got reading, writing files working, at least for json. what other functions should I implement, confirm working prior to merging this?


with testing, I'm still using the dockerfile, bringing both containers up with the compose file doesn't seem to work, but it's most likely a misconfiguration of hadoop or podman. Traffic goes in to the container (write request are fulfilled), but responses don't make it back out.

Given that, should I just document how to set up the dockerfile?

linhr reviewed (commented):
This is an exciting progress!

I just triggered the workflow. Let's format the code using the command here: <https://docs.lakesail.com/sail/main/development/build/rust.html>

Reading and writing JSON files is a perfect scope for this PR. Let's add a FIXME comment in `compose.yml` for future investigation of the container issues, but there is no need to add documentation for manual container setup.

I have a few comments but overall this is quite an impressive change! I can see especially that the Dockerfile requires non-trivial efforts. I think the PR is pretty close to the ready state.

skewballfox commented on scripts/hadoop/Dockerfile:
```diff
@@ -0,0 +1,77 @@
+FROM --platform=linux/amd64 fedora:40
```
This, along with the generating the script in the dockerfile were changes I made when attempting to launch the hdfs instance from compose. I kept getting "EXEC format error" for the command when copying the script into the container and executing it via compose or start

Looking through the commit history, I think it might be due to an empty line above the shebang in `hdfs-init.sh`. I'll move the script back out

skewballfox commented on compose.yml:
```diff
@@ -18,6 +18,29 @@ services:
+      - "9870:9870"
+      - "9000:9000"
+      - "9864:9864"
+    stop_grace_period: 90s
```
Probably not no, this was me trying to figure out why the container was quietly exiting without realizing I needed to end the start script in an endless loop to keep it running

skewballfox commented on crates/sail-plan/src/resolver/data_source.rs:
```diff
@@ -31,8 +32,15 @@ impl PlanResolver<'_> {
-                        url
+
+                    if store.head(&url.prefix()).await.is_ok()  {
+                        //hdfs returns okay if directory exists without trailing delimiter, but returning that causes an error when
+                        //listing files in the directory, hence the double check
```
btw, hdfs-native-object-store just merged a commit that will now throw an error if using head on a directory, so this extra check can probably be removed on their next release

skewballfox commented on scripts/hadoop/Dockerfile:
```diff
@@ -0,0 +1,77 @@
+    </property>
+</configuration>
+EOF
+cat > /usr/bin/hdfs-init.sh <<EOF
```
yeah I didn't realize that was necessary to keep the container running

skewballfox commented:
btw, compose might work on other systems(with or without sudo).  I think part of the issue might be differences in rootless network configuration between `podman-compose` and podmans default on fedora 40. I know when launching hdfs via compose write request go through, because the files are created, outbound traffic is just broken

linhr commented on crates/sail-plan/src/resolver/data_source.rs:
```diff
@@ -31,8 +32,28 @@ impl PlanResolver<'_> {
+
                     if store.head(url.prefix()).await.is_ok() {
-                        url
+                        // HDFS returns okay if directory exists without trailing delimiter, but returning that causes an error when
+                        // listing files in the directory, hence the double check
```
Could you link to the PR in the `hdfs-native-object-store` repository? Let's mention that the logic will no longer be needed once the fix is released in the dependency.

skewballfox replied:
Looks like they pushed a new version to crates.io yesterday which included the change. I removed the check and tested to confirm. It works as expected now.

linhr reviewed (approved):
Thanks for addressing the comments! Also thanks for the notes regarding container setup.

I have one more minor comment, but I feel this is in a good shape!

It seems there is a formatting issue in `compose.yml`. Could you use the following commands to format the non-Rust code? More information can be found [here](https://docs.lakesail.com/sail/main/development/documentation/).

```bash
pnpm install
pnpm run format
pnpm run lint
```

Also, when you commit the changes this time, you can add `[spark tests]` in the commit message so that we can get the Spark tests triggered. After this I think this PR is ready to be merged!

skewballfox commented:
sorry I forgot to put `[spark tests]` in the commit message, could you retrigger it on your end?

linhr commented:
> sorry I forgot to put `[spark tests]` in the commit message, could you retrigger it on your end?

No worries! Since this PR does not change PySpark logic, it's fine to skip the tests. The tests will be run after the code is merged to `main`.

I've also created an issue (#227) so that the tests can be triggered without the special commit message.

linhr reviewed (approved):
@skewballfox We are so glad to have you as the first community contributor of Sail! Thank you for your contribution!
